File: BitManipulation/LeetCode/Medium/PowerSet/PowerSet.cpp
// https://leetcode.com/problems/subsets/
#include<bits/stdc++.h>
using namespace std;
vector<vector<int>> subsets(vector<int> &nums)
{
    vector<vector<int>> allSubs;
    for (int i = 0; i < (1 << nums.size()); i++)
    {
        vector<int> sub;
        for (int j = 0; j < nums.size(); j++)
        {
            if ((1 << j) & i)
            {
                sub.push_back(nums[j]);
            }
        }
        allSubs.push_back(sub);
    }
    return allSubs;
}
int main(void) {
    vector<int> list = {1,2,3};
    for(vector<int> sub: subsets(list)) {
        cout << "[ ";
        for(int ele: sub) {
            cout << ele << " ";
        }
        cout << "]\n";
    }
    return 0;
}

File: BitManipulation/LeetCode/Medium/PowerSet/PowerSet.java
// https://leetcode.com/problems/subsets/

import java.util.ArrayList;
import java.util.List;

public class PowerSet {
    public static List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> allSubs = new ArrayList<>();
        for(int i=0; i < (1 << nums.length); i++) {
            List<Integer> sub = new ArrayList<>();
            for(int j=0; j<nums.length; j++) {
                if(((1 << j) & i) != 0) {
                    sub.add(nums[j]);
                }
            }
            allSubs.add(sub);
        }
        return allSubs;
    }
    public static void main(String[] args) {
        int[] list = {1,2,3};
        System.out.println(subsets(list));
    }
}
